Objects/floatobject.c

```diff
@@ -923,6 +923,52 @@ float___trunc___impl(PyObject *self)
     return PyLong_FromDouble(wholepart);
 }
 
+/*[clinic input]
+float.__floor__
+
+Return the floor of x as an Integral.
+[clinic start generated code]*/
+
+static PyObject *
+float___floor___impl(PyObject *self)
+/*[clinic end generated code: output=e0551dbaea8c01d1 input=38c8fb1dd0b579b4]*/
+{
+    double x = PyFloat_AsDouble(self);
+    double wholepart;
+
+    (void)modf(x, &wholepart);
+    if (LONG_MIN < wholepart && wholepart < LONG_MAX && x < 0) {
+        long aslong = (long)wholepart;
+        if (wholepart != x)
+            --aslong;
+        return PyLong_FromLong(aslong);
+    }
```

Comment:
Why not using floor() and ceil()?

Reply:
According to math.floor implementation, you need to explicitly handle more cases: NaN, infinite, etc. See math_1_to_whatever() in mathmodule.c.

Reply:
should i directly use it or re-implement it in here?

Reply:
The special cases might not need any extra handling if you simply use the `libm` floor followed by `PyLong_FromDouble`.

I suggest removing the fast path for simplicity, and relying on the fast path in `PyLong_FromDouble`. (See #15611.)

Reply:
This has the same problem as `ceil`, for negative numbers: there's no decrement for non-integer negative numbers that don't end up in the fast path.